subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c

```diff
@@ -100,11 +97,27 @@ static void ipm_callback(const struct device *dev, void *context, uint32_t id, v
 	k_work_submit_to_queue(&ctx->ipm_work_q, &ctx->ipm_work);
 }
 
+static inline unsigned int get_instance(struct rpmsg_mi_ctx_shm_cfg *shm)
+{
+	unsigned int this_instance;
+	struct k_spinlock shm_lock;
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&shm_lock);
+	this_instance = shm->instance;
+	shm->instance++;
```

Comment:
The instance variable is incremented in the get_instance function.  If shared memory initialization fails, it will have an incorrect value. Isn't that a problem?

Reply:
A problem yes, but not a big one. What happens is that the current slot in the shared memory region is "skipped" (assuming that you have a shared memory region big enough to host several "slots").

The problem with this issue is that fixing this is not trivial (or at least I'm not seeing any trivial solution). Let me explain why.

One easy solution could be creating an exit / failure code path that would simply `shm->instance--` if for some reason the init fails at any time, to restore the original `shm->instance` value. The problem with this solution is that this reintroduces a race condition when another thread is running the same init function between the `get_instance()` function and the exit code path (because it would still see the old wrong value before it is reset back to the original value).

So to avoid also this race condition we should have the whole init function (or at least the part of it using `shm->instance`) guarded by a spinlock. That is probably not liked by @doki-nordic but I think it is a good enough solution.

Otherwise we should leave the code as is and let the user deal with this kind of problems since he is notified anyway by the error returned by the init function.

Reply:
Was thinking that I could use a mutex instead of a spinlock for the initialization. I'll submit a new revision asap.